File: java/test/src/se/pitch/oss/fedpro/client/hla/TestRTIambassadorClientHla4.java
/*
 *  Copyright (C) 2022 Pitch Technologies AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package se.pitch.oss.fedpro.client.hla;

import hla.rti1516_2025.RtiConfiguration;

import hla.rti1516_2025.fedpro.CallResponse;
import hla.rti1516_2025.fedpro.ExceptionData;
import hla.rti1516_2025.exceptions.RTIexception;
import org.junit.Test;
import se.pitch.oss.fedpro.client_common.exceptions.FedProRtiException;

import java.util.ArrayList;
import java.util.Arrays;

import static org.junit.Assert.assertEquals;
import static se.pitch.oss.fedpro.client.hla.RTIambassadorClientHla4Base.parseRtiConfiguration;
import static se.pitch.oss.fedpro.client_common.RTIambassadorClientGenericBase.extractAndRemoveClientSettings;

public class TestRTIambassadorClientHla4 {

   @Test
   public void clientSettingsAreExtractedAndRemovedCorrectly_When_crcAddressIsProvided()
   {
      // Given
      ArrayList<String> inputList = new ArrayList<>(Arrays.asList(
            "FedPro.connect.hostname=localhost",
            "crcAddress=123.456.789:1234",
            "crcHost=rti.test.local",
            "LRC.skipVersionCheck=false"));

      // When
      String settingsLine = extractAndRemoveClientSettings(inputList, false);

      // Then
      assertEquals(
            settingsLine,
            "connect.hostname=localhost");
      assertEquals(
            inputList,
            Arrays.asList("crcAddress=123.456.789:1234", "crcHost=rti.test.local", "LRC.skipVersionCheck=false"));
   }

   @Test
   public void rtiConfigurationParseCorrectly_When_NoSettingsAreProvided()
   {
      // Given
      RtiConfiguration providedConfiguration = RtiConfiguration.createConfiguration();
      ArrayList<String> inputList = new ArrayList<>();

      // When
      RtiConfiguration rtiConfiguration = parseRtiConfiguration(providedConfiguration, inputList);

      // Then
      assertEquals(rtiConfiguration.rtiAddress(), "");
      assertEquals(rtiConfiguration.configurationName(), "");
      assertEquals(rtiConfiguration.additionalSettings(), "");
   }

   @Test
   public void rtiConfigurationParseCorrectly_When_OnlySettingsAreProvided()
   {
      // Given
      RtiConfiguration providedConfiguration = RtiConfiguration.createConfiguration();
      ArrayList<String> inputList = new ArrayList<>(Arrays.asList(
            "FedPro.connect.timeout=10,",
            "LRC.skipVersionCheck=false"));

      // When
      RtiConfiguration rtiConfiguration = parseRtiConfiguration(providedConfiguration, inputList);

      // Then
      assertEquals(rtiConfiguration.rtiAddress(), "");
      assertEquals(rtiConfiguration.configurationName(), "");
      assertEquals(rtiConfiguration.additionalSettings(), String.join("\n", "FedPro.connect.timeout=10,",
            "LRC.skipVersionCheck=false"));
   }

   @Test
   public void rtiConfigurationParseCorrectly_When_OnlyConfigNameIsProvided()
   {
      // Given
      RtiConfiguration providedConfiguration =
            RtiConfiguration.createConfiguration().withConfigurationName("iosConfiguration");
      ArrayList<String> inputList = new ArrayList<>();

      // When
      RtiConfiguration rtiConfiguration = parseRtiConfiguration(providedConfiguration, inputList);

      // Then
      assertEquals(rtiConfiguration.rtiAddress(), "");
      assertEquals(rtiConfiguration.configurationName(), "iosConfiguration");
      assertEquals(rtiConfiguration.additionalSettings(), "");
   }

   @Test
   public void rtiConfigurationParseCorrectly_When_ConfigNameAndSettingsAreProvided()
   {
      // Given
      RtiConfiguration providedConfiguration =
            RtiConfiguration.createConfiguration().withConfigurationName("iosConfiguration");
      ArrayList<String> inputList = new ArrayList<>(Arrays.asList(
            "FedPro.connect.timeout=10,",
            "LRC.skipVersionCheck=false",
            "FedPro.connect.hostname=localhost"));

      // When
      RtiConfiguration rtiConfiguration = parseRtiConfiguration(providedConfiguration, inputList);

      // Then
      assertEquals(rtiConfiguration.rtiAddress(), "");
      assertEquals(rtiConfiguration.configurationName(), "iosConfiguration");
      assertEquals(rtiConfiguration.additionalSettings(), String.join("\n", "FedPro.connect.timeout=10,",
            "LRC.skipVersionCheck=false",
            "FedPro.connect.hostname=localhost"));
   }

   @Test(expected = RTIexception.class)
   public void throwOnException_throw_RTIexception()
   throws FedProRtiException
   {
      ClientConverter clientConverter = new ClientConverter();
      RTIambassadorClientHla4Base ambassadorClient = new RTIambassadorClientHla4Base(clientConverter);

      // Given
      String message = "Placeholder message";
      CallResponse exceptionCallResponse = CallResponse.newBuilder()
            .setExceptionData(ExceptionData.newBuilder().setExceptionName("RTIexception").setDetails(message))
            .build();

      // When
      ambassadorClient.throwOnException(exceptionCallResponse);

      // Then throw
   }

}

File: cpp/test/services-2025/TestRTIAmbassadorClientHla4.cpp
/***********************************************************************
  Copyright (C) 2023 Pitch Technologies AB

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 **********************************************************************/

// Silence clang-tidy issues reported for gtest macros.
// NOLINTBEGIN(cppcoreguidelines-avoid-non-const-global-variables)

#include "services-hla4/RTIambassadorClientAdapter.h"
#include "services-common/RTIambassadorClientGenericBase.h"
#include "services-common/RtiConfiguration.h"
#include "services-common/SettingsParser.h"

#include <gtest/gtest.h>

using namespace FedPro;
using namespace RTI_NAMESPACE;

// TestAmbassador acts as RTIambassador test fixture for HLA 4.
class TestRTIambassadorClientHla4 : public ::testing::Test, public RTIambassadorClientAdapter
{
public:
   TestRTIambassadorClientHla4() = default;
};

TEST(TestRTIambassadorClientHla4, clientSettingsAreExtractedAndRemovedCorrectly_When_crcAddressIsProvided)
{
   // Given
   std::vector<std::wstring> inputList
         {L"FedPro.connect.hostname=localhost", L"crcAddress=123.456.789:1234", L"crcHost=rti.test.local",
          L"LRC.skipVersionCheck=false"};

   // When
   std::string settingsLine = RTIambassadorClientGenericBase::extractAndRemoveClientSettings(inputList, false);

   // Then
   EXPECT_EQ(settingsLine, "connect.hostname=localhost");
   EXPECT_EQ(inputList,
             (std::vector<std::wstring>{L"crcAddress=123.456.789:1234", L"crcHost=rti.test.local",
                                        L"LRC.skipVersionCheck=false"}));
}

TEST(TestRTIambassadorClientHla4, rtiConfigurationParseCorrectly_When_NoSettingsAreProvided)
{
   // Given
   RtiConfiguration providedConfiguration = RtiConfiguration::createConfiguration();
   std::vector<std::wstring> inputList;

   // When
   RtiConfiguration
         rtiConfiguration = RTIambassadorClientAdapter::parseRtiConfiguration(providedConfiguration, inputList);

   // Then
   EXPECT_EQ(rtiConfiguration.rtiAddress(), L"");
   EXPECT_EQ(rtiConfiguration.configurationName(), L"");
   EXPECT_EQ(rtiConfiguration.additionalSettings(), L"");
}

TEST(TestRTIambassadorClientHla4, rtiConfigurationParseCorrectly_When_OnlySettingsAreProvided)
{
   // Given
   RtiConfiguration providedConfiguration = RtiConfiguration::createConfiguration();
   std::vector<std::wstring> inputList = {L"FedPro.connect.timeout=10,", L"LRC.skipVersionCheck=false"};

   // When
   RtiConfiguration
         rtiConfiguration = RTIambassadorClientAdapter::parseRtiConfiguration(providedConfiguration, inputList);

   // Then
   EXPECT_EQ(rtiConfiguration.rtiAddress(), L"");
   EXPECT_EQ(rtiConfiguration.configurationName(), L"");
   EXPECT_EQ(rtiConfiguration.additionalSettings(), L"FedPro.connect.timeout=10,\nLRC.skipVersionCheck=false");
}

TEST(TestRTIambassadorClientHla4, rtiConfigurationParseCorrectly_When_OnlyConfigNameIsProvided)
{
   // Given
   RtiConfiguration
         providedConfiguration = RtiConfiguration::createConfiguration().withConfigurationName(L"iosConfiguration");
   std::vector<std::wstring> inputList;

   // When
   RtiConfiguration
         rtiConfiguration = RTIambassadorClientAdapter::parseRtiConfiguration(providedConfiguration, inputList);

   // Then
   EXPECT_EQ(rtiConfiguration.rtiAddress(), L"");
   EXPECT_EQ(rtiConfiguration.configurationName(), L"iosConfiguration");
   EXPECT_EQ(rtiConfiguration.additionalSettings(), L"");
}

TEST(TestRTIambassadorClientHla4, rtiConfigurationParseCorrectly_When_ConfigNameAndSettingsAreProvided)
{
   // Given
   RtiConfiguration
         providedConfiguration = RtiConfiguration::createConfiguration().withConfigurationName(L"iosConfiguration");
   std::vector<std::wstring> inputList =
         {L"FedPro.connect.timeout=10,", L"LRC.skipVersionCheck=false", L"FedPro.connect.hostname=localhost"};

   // When
   RtiConfiguration
         rtiConfiguration = RTIambassadorClientAdapter::parseRtiConfiguration(providedConfiguration, inputList);

   // Then
   EXPECT_EQ(rtiConfiguration.rtiAddress(), L"");
   EXPECT_EQ(rtiConfiguration.configurationName(), L"iosConfiguration");
   EXPECT_EQ(rtiConfiguration.additionalSettings(),
             L"FedPro.connect.timeout=10,\nLRC.skipVersionCheck=false\nFedPro.connect.hostname=localhost");
}

// NOLINTEND(cppcoreguidelines-avoid-non-const-global-variables)
